### PR TITLE
Keep the buffer in the invalid UTF8 handling path of String(unsafeUninitializedCapacity:initializingWith:) alive while we're using it

### PR DESCRIPTION
**Pull request by Catfish-Man**

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
Looks like all noise for perf, but someone should figure out what's up with DictionaryOfAnyHashableStrings_insert, it's being weird lately

File: stdlib/public/core/StringCreate.swift [hunk 1]
```diff
@@ -120,8 +120,15 @@ extension String {
       )
       return result.asString
     case .error(let initialRange):
-      //This could be optimized to use excess tail capacity
-      return repairUTF8(result.codeUnits, firstKnownBrokenRange: initialRange)
+      // repairUTF8 will call reserveCapacity, invalidating result.codeUnits if
+      // we use it directly. This could be a little faster, but this error path
+      // shouldn't be perf-sensitive anyway.
+      let temporaryCodeUnits = Array(result.codeUnits)
+      defer { _fixLifetime(temporaryCodeUnits) }
```
Comment (stephentyrone) [inline stdlib/public/core/StringCreate.swift:127]: Why do we need _fixLifetime here?

File: test/stdlib/StringCreate.swift [hunk 2]
```diff
@@ -57,22 +57,33 @@ if #available(macOS 10.16, iOS 14.0, watchOS 7.0, tvOS 14.0, *) {
       }
       expectEqual(expected, actual)
     }
-
+
```
Comment (glessard) [inline test/stdlib/StringCreate.swift:60]: Nit: trailing whitespace.

File: stdlib/public/core/StringCreate.swift (hunk 1 above)
Comment (Catfish-Man) [reply stdlib/public/core/StringCreate.swift:127]: Ah, right you are, it does that internally

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
UGH I reverted the test changes accidentally. Ok, I'm going to eat breakfast first so I can think properly, THEN fix this PR.

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Review by glessard — APPROVED** [review]

**Review by milseman — APPROVED** [review]
LGTM

**Pull request merged**